flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala

```diff
@@ -0,0 +1,463 @@
+  private def instantiateDeserializationFields(): Array[Any] = {
+    val newFields = new Array[Any](serializers.length)
+    var i = 0
+    while (i < serializers.length) {
+      newFields(i) = serializers(i).createInstance()
+      i += 1
+    }
+    newFields
+  }
+
+  // --------------------------------------------------------------------------------------------
+  //  Comparator Methods
+  // --------------------------------------------------------------------------------------------
+
+  override def getFlatComparator(flatComparators: util.List[TypeComparator[_]]): Unit =
+    comparators.foreach {
+      case ctc: CompositeTypeComparator[_] => ctc.getFlatComparator(flatComparators)
+      case c@_ => flatComparators.add(c)
+    }
+
+  override def compareToReference(referencedComparator: TypeComparator[Row]): Int = {
+    val other: RowComparator = referencedComparator.asInstanceOf[RowComparator]
+    var i = 0
+    try {
+      while (i < keyPositions.length) {
+        val comparator = comparators(i).asInstanceOf[TypeComparator[Any]]
+        val otherComparator = other.comparators(i).asInstanceOf[TypeComparator[Any]]
+
+        val nullCheck = comparator.equalToReference(nullChecker)
+        val nullCheckOther = otherComparator.equalToReference(nullChecker)
+
+        var cmp = 0
+        // both values are null -> equality
+        if (nullCheck && nullCheckOther) {
+          cmp = 0
+        }
+        // one value is null -> inequality
+        // order is considered for basic types
+        else if (nullCheck || nullCheckOther) {
+          if (comparator.isInstanceOf[BasicTypeComparator[_]]) {
+            val basicComp = comparator.asInstanceOf[BasicTypeComparator[_]]
+            if (nullCheck) {
+              return if (basicComp.isAscendingComparison) 1 else -1
+            }
+            else if (nullCheckOther) {
+              return if (basicComp.isAscendingComparison) -1 else 1
+            }
+          }
+          else {
+            return if (nullCheck) 1 else -1
+          }
+        }
+        else {
+          cmp = comparator.compareToReference(otherComparator)
+        }
+
+        if (cmp != 0) {
+          return cmp
+        }
+        i = i + 1
+      }
+      0
+    }
+    catch {
+      case npex: NullPointerException =>
+        throw new NullKeyFieldException(keyPositions(i))
+      case iobex: IndexOutOfBoundsException =>
+        throw new KeyFieldOutOfBoundsException(keyPositions(i))
+    }
+  }
+
+  override def compareSerialized(firstSource: DataInputView, secondSource: DataInputView): Int = {
+    val len = serializers.length
+    val keyLen = keyPositions.length
+
+    readIntoNullMask(len, firstSource, nullMask1)
+    readIntoNullMask(len, secondSource, nullMask2)
+
+    // deserialize
+    var i = 0
+    while (i < len) {
+      val serializer = serializers(i).asInstanceOf[TypeSerializer[Any]]
+
+      // deserialize field 1
+      if (nullMask1(i)) {
+        deserializedKeyFields1(i) = null
```

Comment:
If you use `nullMask1` during the comparison step to check if a field `null` is, you don't have to set the field to `null` here and read the object the next time without a `reuse` object.